src/csharp/Microsoft.Spark.UnitTest/UdfSerDeTests.cs

```diff
@@ -0,0 +1,152 @@
+                var that = obj as TestClass;
+
+                if (that == null)
+                {
+                    return false;
+                }
+
+                return _str == that._str;
+            }
+
+            public override int GetHashCode()
+            {
+                return base.GetHashCode();
+            }
+        }
+
+        [Fact]
+        public void TestUdfSerDe()
+        {
+            {
+                // Without closure.
+                Func<int, int> udf = i => 10 * i;
+                VerifyUdfSerDe(udf, false);
+            }
+
+            {
+                // With closure where the delegate target is an anonymous class.
+                // The target will contain fields ["tc1", "tc2"], where "tc1" is
+                // non null and "tc2" is null.
+                TestClass tc1 = new TestClass("Test");
+                TestClass tc2 = null;
+                Func<string, string> udf =
+                    (s) =>
+                    {
+                        if (tc2 == null)
+                        {
+                            return tc1.Concat(s);
+                        }
+                        return string.Empty;
+                    };
+                VerifyUdfSerDe(udf, true);
+            }
+
+            {
+                // With closure where the delegate target is TestClass
+                // and target's field "_str" is set to "Test".
+                TestClass tc = new TestClass("Test");
+                Func<string, string> udf = tc.Concat;
+                VerifyUdfSerDe(udf, true);
+            }
+
+            {
+                // With closure where the delegate target is TestClass,
+                // and target's field "_str" is set to null.
+                TestClass tc = new TestClass(null);
+                Func<string, string> udf = tc.Concat;
+                VerifyUdfSerDe(udf, true);
+            }
+        }
+
+        private void VerifyUdfSerDe(Delegate udf, bool hasClosure)
+        {
+            byte[] serializedUdf = SerializeUdf(udf);
+            DeserializeAndVerify(udf, serializedUdf, hasClosure);
+        }
+
+        private byte[] SerializeUdf(Delegate udf)
+        {
+            UdfSerDe.UdfData udfData = UdfSerDe.Serialize(udf);
+
+            using (var ms = new MemoryStream())
+            {
+                var bf = new BinaryFormatter();
+                bf.Serialize(ms, udfData);
+                return ms.ToArray();
+            }
+        }
+
+        private void DeserializeAndVerify(
+            Delegate udf,
+            byte[] serializedUdf,
+            bool hasClosure)
+        {
+            using (var ms = new MemoryStream(serializedUdf, false))
+            {
+                var bf = new BinaryFormatter();
+                var udfData = (UdfSerDe.UdfData)bf.Deserialize(ms);
+                VerifyUdfData(UdfSerDe.Serialize(udf), udfData, hasClosure);
+
+                Delegate deserializedUdf = UdfSerDe.Deserialize(udfData);
```

Comment:
I think we need `VerifyDelegate()` and `VerifyUdfData` separately to be clear.

Reply:
Added VerifyDelegate